### DISCUSSION TITLE
[Firebase] Relation DB vs or NOT(NoSQL)

### DISCUSSION THREAD

**Discussion by yudonlee** [DeveloperAcademy-POSTECH/MacC-Team-Vegeting#25] [자료공유]
# 1. Firebase의 원칙 
모임, 채팅, 유저 세가지를 연결할 수 있는 서버 구조 구현
파이어베이스는 NoSQL구조를 가지기 때문에, 필요한 값은 각자 다른 Document영역에 중복돼서 가질 수 있다는 대원칙을 전제로 합니다.(해당 부분에 대한 레퍼런스는 출처를 확인하는 걸 추천합니다, 또한 Relational로 설계했던 이전 MC3프로젝트들에서 서버 request증가등 부작용이 있었기 떄문입니다.)

# 2. 두가지의 차이점을 예시를 갖고 비교
Relation DB와 NoSQL 구조에 대한 차이점을 우리 프로젝트의 예시를 통해 설명하겠습니다.
## Relational DB
- 채팅방에 User들의 프로필 사진이 나타나야합니다.
- 우리는 UserID를 통해 User Document를 서버에 요청하여 ImageURL을 받아옵니다
- imageURL을 통해서 서버에 image를 요청합니다

## 중복 저장(NoSQL)
- 채팅방에 User들의 프로필 사진이 나타나야합니다.
- 채팅방 Document에 저장된 User의 imageURL을 가지고 서버에 요청합니다.

### 장점 
결국 중복 저장 구조를 가졌을 때 서버 접속 Depth를 줄여주기 때문에 해당 원칙을 채택했습니다.
### 단점
부가적인 저장용량이 조금은 늘어난다는 단점이 있습니다.
Data Integrity를 위배 할 수 있습니다.(서로 연결된 모든 field의 데이터가 업데이트가 되어야 합니다. 하지만 Relation DB또한 똑같은 문제를 갖고 있으며 두개다 완벽한 보장이 안된다는 점은 마찬가지 입니다.)

![서버구조](https://user-images.githubusercontent.com/39371835/197378081-10befb59-9310-45a7-b098-55091c006217.jpg)

# Reference
[공식 Google Firebase에서 Storage DB Modeling 설명](https://www.youtube.com/watch?v=v_hR4K4auoQ)

**Comment by wonhui-kim** [comment]
NoSQL 방식이 저장 용량이 늘어난다는 단점이 있지만 현재 앱의 규모가 크지 않고, 요청 시간이 줄어들어서 사용자에게 더 편리한 경험을 줄 수 있어서 더 좋은 방식이라고 생각합니다!

**Reply by ccdkk** [reply]
자세하게 정리해주셔서 이해가 잘되었습니다! 저도 NoSQL방식이 서버의 부하를 줄일 수 있어서 더 좋은 것 같습니다!
